### PR TITLE
FIX APGD implementation

### LINKED ISSUE #60: Shouldn't you have two sequences of variables for the accelerated version?
For instance this line only has variables _w_

https://github.com/benchopt/benchOpt/blob/01acb156bd76708b01a30c3de8918609f1cde3ff/benchmarks/lasso/solvers/python_pgd.py#L31

but I think you need two sequences. Something like this (note the additional variables in x)

       L = np.linalg.norm(self.X, ord=2) ** 2
       n_features = self.X.shape[1]
       w = np.zeros(n_features)
       x = np.zeros(n_features)
       t_new = 1

        for _ in range(n_iter):
            if self.use_acceleration:
                t_old = t_new
                t_new = (1 + np.sqrt(1 + 4 * t_old ** 2)) / 2
            grad = self.X.T.dot(self.X.dot(w) - self.y)
            x_old = x.copy()
            x -= grad / L
            x = self.st(x, self.lmbd / L)  # Not sure what this step does
            if self.use_acceleration:
                w = x+ (t_old - 1.) / t_new * (x -x_old)
        self.w = w

### PR DESCRIPTION
**Pull request by mathurinm**
Comparison of the 2 implems:
image

closes #60

### CODE REVIEW

**Review by agramfort — APPROVED** [review]

File: benchmarks/lasso/solvers/python_pgd.py [hunk 1]
```diff
@@ -18,17 +18,20 @@ def run(self, n_iter):
...
         for _ in range(n_iter):
             if self.use_acceleration:
                 t_old = t_new
                 t_new = (1 + np.sqrt(1 + 4 * t_old ** 2)) / 2
                 w_old = w.copy()
-            grad = self.X.T.dot(self.X.dot(w) - self.y)
-            w -= grad / L
+            else:
+                z = w
+            w -= self.X.T @ (self.X @ z - self.y) / L
```
Comment (TomDLT) [inline benchmarks/lasso/solvers/python_pgd.py:30]:
```suggestion
            w = z - self.X.T @ (self.X @ z - self.y) / L
```

**Review by TomDLT — CHANGES_REQUESTED** [review]
That explains the poor performances of the old APGD on the Boston dataset:
Figure_1
However, your fix diverges:
Figure_1b
This is because the gradient step should be taken from the auxiliary variable. With my proposed fix:
Figure_2

**Comment by mathurinm** [comment]
Thanks for the thorough review !
It is still weird that pgd outperforms apgd on Boston, don't you think ?

**Comment by TomDLT** [comment]
Rerunning the benchmark with 100 repetitions.
Figure_1
Figure_2
Figure_3

**Comment by agramfort** [comment]
I also get now some:

|----Python-PGD[use_acceleration=False]: done (did not converge)
|----Python-PGD[use_acceleration=True]: done (did not converge)

on simulated or boston.

it seems we have a regression here.

**Review by agramfort — APPROVED** [review]
@TomDLT I changed a bit more to avoid an extra copy for PDG without accel.

**Comment by agramfort** [comment]
and forget about my brain fart about regression above. I was using -n instead of -r ...

all good on my side

**Review by TomDLT — APPROVED** [review]

**Review by tomMoral — APPROVED** [review]

**Comment by tomMoral** [comment]
Thanks everyone!

**Pull request merged**